File: 3Sum.cpp
// 15 Array medium

#include <vector>
class Solution {
  public:
    vector<vector<int>> threeSum( vector<int> &nums ) {
        vector<vector<int>> ret;
        if ( nums.size() < 3 ) {
            return ret;
        }
        sort( nums.begin(), nums.end() );
        for ( size_t i = 0; i < nums.size() - 2; i++ ) {
            int target = -nums[ i ];
            int front = i + 1;
            int end = nums.size() - 1;
            // 寻找有序数组两数和为target，采用两边收缩遍历法
            while ( end > front ) {
                int sum = nums[ front ] + nums[ end ];
                if ( sum < target ) {
                    front++;
                } else if ( sum > target ) {
                    end--;
                } else {
                    vector<int> triplet( 3 );
                    triplet[ 0 ] = nums[ i ];
                    triplet[ 1 ] = nums[ front ];
                    triplet[ 2 ] = nums[ end ];
                    ret.push_back( triplet );

                    // 因为不要重复的三元组，所以找到一个正确答案后往里收缩时应该去掉重复值
                    // 两边都收缩到更改了值为止
                    while ( front < end && nums[ front ] == triplet[ 1 ] ) {
                        front++;
                    }
                    while ( front < end && nums[ end ] == triplet[ 2 ] ) {
                        end--;
                    }
                }
            }
            // 跳过后面相同的nums[i]，防止重复
            while ( i + 1 < nums.size() && nums[ i + 1 ] == nums[ i ] ) {
                i++;
            }
        }
        return ret;
    }
};

File: 3sum.js
// 15 Array medium

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
  nums.sort((a, b) => a - b)
  const ret = []
  for (let i = 0; i < nums.length; i++) {
    // 找到第一个与左边不同的数字
    while (nums[i] === nums[i-1]) i++
    if (nums[i] > 0) break
    // 从右边的数组中寻找
    let left = i+1, right = nums.length-1
    while (left < right) {
      const sum = nums[i] + nums[left] + nums[right]
      if (sum === 0) {
        ret.push([nums[i], nums[left], nums[right]])
      }
      if (sum < 0) {
        // 如果和偏小，则左指针++，一直到第一个与左边不同的数字
        do {
          left++
        } while (nums[left] === nums[left-1])
      } else {
        do {
          right--
        } while (nums[right] === nums[right+1])
      }
    }
  }
  return ret
}

console.log(threeSum([0,0, 0]))
console.log(threeSum([-2,1, 1]))
console.log(threeSum([-1,0,1,0]))
console.log(threeSum([-4,-1,-1,0,1,2]))